core/src/main/java/org/elasticsearch/transport/TcpTransport.java

```diff
@@ -1384,4 +1439,146 @@ public void onFailure(Exception e) {
             }
         }
     }
+
+    private final LongObjectMap<TransportResponseHandler<?>> pendingHandshakes = new LongObjectHashMap<>();
+    private final AtomicLong requestIdGenerator = new AtomicLong();
+    private final CounterMetric numHandshakes = new CounterMetric();
+    private static final String HANDSHAKE_ACTION_NAME = "internal:tcp/handshake";
+
+
+    protected void handleHandshakeRequest(Channel c, long requestId, Version version) throws IOException {
+        VersionHandshakeResponse response = new VersionHandshakeResponse(getCurrentVersion());
+        try (BytesStreamOutput stream = new BytesStreamOutput()) {
+            stream.setVersion(version);
+            threadPool.getThreadContext().writeTo(stream);
+            response.writeTo(stream);
+            BytesReference bytes = stream.bytes();
+            sendMessage(c, new CompositeBytesReference(buildHandshakeMessage(false, requestId,
+                version, bytes.length()), bytes) , () -> {});
+        }
+    }
+
+    protected void sendHandshakeRequest(Channel c, long requestId) throws IOException {
+        try (BytesStreamOutput stream = new BytesStreamOutput()) {
+            stream.setVersion(getCurrentVersion().minimumCompatibilityVersion());
+            threadPool.getThreadContext().writeTo(stream);
+            stream.writeString(HANDSHAKE_ACTION_NAME);
+            BytesReference bytes = stream.bytes();
+            sendMessage(c, new CompositeBytesReference(buildHandshakeMessage(true, requestId,
+                getCurrentVersion().minimumCompatibilityVersion(), bytes.length()), bytes) , () -> {});
+        }
+    }
+
+    protected BytesReference buildHandshakeMessage(boolean req, long requestId, Version version, int length) throws IOException {
+        byte status = req ? TransportStatus.setRequest(TransportStatus.setHandshake((byte)0))
+            : TransportStatus.setResponse(TransportStatus.setHandshake((byte)0));
+        return buildHeader(requestId, status, version, length);
+    }
+
+    private static final class VersionHandshakeResponse extends TransportResponse {
+        private Version version;
+
+        private VersionHandshakeResponse(Version version) {
+            this.version = version;
+        }
+
+        private VersionHandshakeResponse() {
+        }
+
+        @Override
+        public void readFrom(StreamInput in) throws IOException {
+            super.readFrom(in);
+            version = Version.readVersion(in);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            Version.writeVersion(version, out);
```

Comment:
Can you assert here `version != null`? The empty ctor is only for doing readFrom right?